src/Domain/LeanCode.DomainModels.Generators/IdSource.cs

```diff
@@ -140,6 +144,114 @@ public static bool IsValid([NotNullWhen(true)] string? v)
+    using global::LeanCode.DomainModels.Ids;
+    using global::LeanCode.DomainModels.Ulids;
+    #pragma warning restore CS8019
+
+    [JsonConverter(typeof(StringTypedIdConverter<{{data.TypeName}}>))]
+    [DebuggerDisplay("{Value}")]
+    public readonly partial record struct {{data.TypeName}} : IPrefixedTypedId<{{data.TypeName}}>
+    {
+        private const int ValueLength = {{valueLength}};
+        private const char Separator = '_';
+        private const string TypePrefix = "{{prefix}}";
+
+        public static int RawLength { get; } = {{valueLength + 1 + prefix.Length}};
+        public static readonly {{data.TypeName}} Empty = new(Ulid.Empty);
+
+        private readonly string? value;
+
+        public string Value => value ?? Empty.Value;
+        public bool IsEmpty => value is null || value == Empty;
+        public Ulid Ulid => value is null ? Ulid.Empty : Ulid.Parse(value.AsSpan()[{{prefix.Length + 1}}..]);
+
+        private {{data.TypeName}}(string v) => value = v;
+
+        public {{data.TypeName}}(Ulid v) => value = string.Create(null, stackalloc char[RawLength], $"{TypePrefix}{Separator}{v}");
+
+        public static {{data.TypeName}} New() => new(Ulid.NewUlid());
+
+        public static {{data.TypeName}} Parse(string? v)
+        {
+            if (TryDeconstruct(v.AsSpan(), out var ulid))
+            {
+                return new {{data.TypeName}}(ulid);
+            }
+            else
+            {
+                throw new FormatException(
+                    $"The ID has invalid format. It should look like {TypePrefix}{Separator}(id value)."
+                );
+            }
+        }
+
+        [return: NotNullIfNotNull("id")]
+        public static {{data.TypeName}}? ParseNullable(string? id) => id is string v ? Parse(v) : ({{data.TypeName}}?)null;
+
+        public static bool TryParse([NotNullWhen(true)] string? v, out {{data.TypeName}} id)
+        {
+            if (TryDeconstruct(v, out var ulid))
+            {
+                id = new {{data.TypeName}}(ulid);
+                return true;
+            }
+            else
+            {
+                id = default;
+                return false;
+            }
+        }
+
+        public static bool TryDeconstruct(ReadOnlySpan<char> span, out Ulid rawUlid)
+        {
+            rawUlid = Ulid.Empty;
+
+            return span.Length == RawLength
+                && span.StartsWith(TypePrefix)
+                && span[{{prefix.Length}}] == Separator
+                && Ulid.TryParse(span[{{prefix.Length + 1}}..], out rawUlid);
+        }
+
+        public static bool IsValid([NotNullWhen(true)] string? v)
+        {
+            return TryDeconstruct(v.AsSpan(), out _);
+        }
+
+        public bool Equals({{data.TypeName}} other) => Value.Equals(other.Value, StringComparison.Ordinal);
+        public int CompareTo({{data.TypeName}} other) => string.Compare(Value, other.Value, StringComparison.Ordinal);
```

Comment:
I guess you could compare just the `Ulid` parts here, which might've been more efficient.

Reply:
Nah, string comparison is vectorized anyway, little to zero gain but a readability cost